### PR TITLE
Extract templates from jar directly rather than getResource() if possible

### LINKED ISSUE #365: Sporadic IOExceptions during parsedTemplateForResource
We're seeing sporadic failures during builds from AutoValue during `parsedTemplateForResource`. We're really at a loss for what's going on, and are happy to dig in to anything you want us to try. Using AutoValue 1.3-rc2, one internal custom extension being used in Android projects. They look like this (both with gradle and buck builds):

```
Caused by: java.lang.AssertionError: java.io.IOException: Stream closed

	at com.google.auto.value.processor.TemplateVars.parsedTemplateForResource(TemplateVars.java:111)
	at com.google.auto.value.processor.AutoValueTemplateVars.<clinit>(AutoValueTemplateVars.java:184)
	at com.google.auto.value.processor.AutoValueProcessor.processType(AutoValueProcessor.java:441)
	at com.google.auto.value.processor.AutoValueProcessor.process(AutoValueProcessor.java:150)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	... 34 more
Caused by: java.io.IOException: Stream closed

	at com.google.auto.value.processor.escapevelocity.Parser.next(Parser.java:141)
	at com.google.auto.value.processor.escapevelocity.Parser.<init>(Parser.java:84)
	at com.google.auto.value.processor.escapevelocity.Template.parseFrom(Template.java:57)
	at com.google.auto.value.processor.TemplateVars.parsedTemplateForResource(TemplateVars.java:107)
	... 45 more
```

```
java.lang.RuntimeException: java.lang.AssertionError: java.io.IOException: Stream closed

	at com.sun.tools.javac.main.Main.compile(Main.java:553)
	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
	at com.facebook.buck.jvm.java.Jsr199Javac.buildWithClasspath(Jsr199Javac.java:247)
	at com.facebook.buck.jvm.java.Jsr199Javac.buildWithClasspath(Jsr199Javac.java:155)
	at com.facebook.buck.jvm.java.JavacStep.tryBuildWithFirstOrderDeps(JavacStep.java:154)
	at com.facebook.buck.jvm.java.JavacStep.execute(JavacStep.java:137)
	at com.facebook.buck.step.DefaultStepRunner.runStepForBuildTarget(DefaultStepRunner.java:63)
	at com.facebook.buck.rules.CachingBuildEngine.executeCommandsNowThatDepsAreBuilt(CachingBuildEngine.java:1223)
	at com.facebook.buck.rules.CachingBuildEngine.access$4(CachingBuildEngine.java:1204)
	at com.facebook.buck.rules.CachingBuildEngine$3$1.call(CachingBuildEngine.java:304)
	at com.facebook.buck.rules.CachingBuildEngine$3$1.call(CachingBuildEngine.java:1)
	at com.facebook.buck.util.concurrent.WeightedListeningExecutorService$1.apply(WeightedListeningExecutorService.java:65)
	at com.facebook.buck.util.concurrent.WeightedListeningExecutorService$1.apply(WeightedListeningExecutorService.java:1)
	at com.google.common.util.concurrent.Futures$AsyncChainingFuture.doTransform(Futures.java:1442)
	at com.google.common.util.concurrent.Futures$AsyncChainingFuture.doTransform(Futures.java:1433)
	at com.google.common.util.concurrent.Futures$AbstractChainingFuture.run(Futures.java:1408)
	at com.google.common.util.concurrent.Futures$2$1.run(Futures.java:1177)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: java.io.IOException: Stream closed

	at com.google.auto.value.processor.TemplateVars.parsedTemplateForResource(TemplateVars.java:111)
	at com.google.auto.value.processor.AutoValueTemplateVars.<clinit>(AutoValueTemplateVars.java:184)
	at com.google.auto.value.processor.AutoValueProcessor.processType(AutoValueProcessor.java:441)
	at com.google.auto.value.processor.AutoValueProcessor.process(AutoValueProcessor.java:150)
	at com.facebook.buck.jvm.java.TracingProcessorWrapper.process(TracingProcessorWrapper.java:104)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
	at com.sun.tools.javac.main.Main.compile(Main.java:523)
	... 20 more
Caused by: java.io.IOException: Stream closed

	at java.util.zip.InflaterInputStream.ensureOpen(InflaterInputStream.java:67)
	at java.util.zip.InflaterInputStream.read(InflaterInputStream.java:142)
	at java.io.FilterInputStream.read(FilterInputStream.java:133)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.read1(BufferedReader.java:210)
	at java.io.BufferedReader.read(BufferedReader.java:286)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.read(BufferedReader.java:182)
	at java.io.LineNumberReader.read(LineNumberReader.java:126)
	at com.google.auto.value.processor.escapevelocity.Parser.next(Parser.java:141)
	at com.google.auto.value.processor.escapevelocity.Parser.<init>(Parser.java:84)
	at com.google.auto.value.processor.escapevelocity.Template.parseFrom(Template.java:57)
	at com.google.auto.value.processor.TemplateVars.parsedTemplateForResource(TemplateVars.java:107)
	... 32 more
```

### PR DESCRIPTION
**Pull request by hzsweers**
Followup from and hopefully helps resolve #365

The one thing about this is I'm not sure how to test this locally, as the classes aren't in an actual jar during tests. I did try manually running tests with a jar location hardcoded, and that did pass. Let me know what you think.

### CODE REVIEW

File: value/src/main/java/com/google/auto/value/processor/TemplateVars.java [hunk 1]
```diff
@@ -97,7 +102,23 @@ String toText() {
...
+    String jar = getJARFromURL(AutoValueTemplateVars.class);
+    InputStream in = null;
+    if (jar != null) {
+      // Prefer fetching from the jar file if available
+      File file = new File(jar);
+      try {
+        JarFile jarFile = new JarFile(file);
+        ZipEntry entry = jarFile.getEntry("com/google/auto/value/processor/".replace('/', File.separatorChar) + resourceName);
+        in = jarFile.getInputStream(entry);
+      } catch (IOException e) {
```
Comment (kageiit) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: e -> ignored

File: value/src/main/java/com/google/auto/value/processor/TemplateVars.java [hunk 2]
```diff
@@ -97,7 +102,23 @@ String toText() {
...
   static Template parsedTemplateForResource(String resourceName) {
-    InputStream in = AutoValueTemplateVars.class.getResourceAsStream(resourceName);
+    String jar = getJARFromURL(AutoValueTemplateVars.class);
+    InputStream in = null;
+    if (jar != null) {
+      // Prefer fetching from the jar file if available
+      File file = new File(jar);
+      try {
+        JarFile jarFile = new JarFile(file);
+        ZipEntry entry = jarFile.getEntry("com/google/auto/value/processor/".replace('/', File.separatorChar) + resourceName);
```
Comment (kageiit) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: check if entry is non null

File: value/src/main/java/com/google/auto/value/processor/TemplateVars.java [hunk 3]
```diff
@@ -97,7 +102,23 @@ String toText() {
...
+        in = jarFile.getInputStream(entry);
+      } catch (IOException e) {
+        // ignore
+      }
+    }
+    if (in == null) {
```
Comment (kageiit) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: else?

Comment (hzsweers) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: handled below

File: value/src/main/java/com/google/auto/value/processor/TemplateVars.java (hunk 1 above)
Comment (hzsweers) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: 59d2b08

File: value/src/main/java/com/google/auto/value/processor/TemplateVars.java (hunk 2 above)
Comment (hzsweers) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: Opting to just have it fallthrough 59d2b08

**Comment by eamonnmcmanus** [comment]
Thanks for putting together this PR! I'd like to follow a slightly different approach, but the PR has at least prompted me to do that. I think it is worth trying to use ClassLoader.getResourceAsStream still, and only falling back to the jar-reading logic if that gets IOException. The reason is that URLClassLoader memory-maps jar files, so I would expect getResource to be much more efficient than reopening the same jar file again and searching for the resource in it.

I have a change out for review within Google and hopefully it will be pushed soon.

File: value/src/main/java/com/google/auto/value/processor/TemplateVars.java [hunk 4]
```diff
@@ -111,6 +132,17 @@ static Template parsedTemplateForResource(String resourceName) {
     }
   }
 
+  private static String getJARFromURL(Class<?> clazz) {
+    URL url = clazz.getResource('/' + clazz.getName().replace('.', '/') + ".class");
```
Comment (eamonnmcmanus) [inline value/src/main/java/com/google/auto/value/processor/TemplateVars.java]: FWIW, there's no guarantee that you can load classes as resource this way, though in practice you can with URLClassLoader.

**Comment by hzsweers** [comment]
Thanks! Looking forward to seeing the change

**Pull request closed without merging**